### ISSUE TITLE
let_underscore_untyped fails to consider configured msrv

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#10410]
### Summary

Clippy's let_underscore_untyped lint triggers on code that is 100% good and idiomatic in projects that target 1.58 or older.

### Lint Name

let_underscore_untyped

### Reproducer

```rust
// src/main.rs

#![deny(clippy::pedantic)]

fn f() -> Result<u8, ()> {
    Err(())
}

fn main() {
    let _ = f();
}
```

```toml
# .clippy.toml

msrv = "1.58.0"
```

```console
$ cargo clippy

error: non-binding `let` without a type annotation
 --> src/main.rs:8:5
  |
8 |     let _ = f();
  |     ^^^^^^^^^^^^
  |
  = help: consider adding a type annotation or removing the `let` keyword
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_untyped
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[deny(clippy::let_underscore_untyped)]` implied by `#[deny(clippy::pedantic)]`
```

Clippy's suggested code does not compile.

```console
error[E0658]: destructuring assignments are unstable
 --> src/main.rs:8:5
  |
8 |     _ = f();
  |     ^
  |
  = note: see issue #71126 <https://github.com/rust-lang/rust/issues/71126> for more information

error[E0658]: destructuring assignments are unstable
 --> src/main.rs:8:7
  |
8 |     _ = f();
  |     - ^
  |     |
  |     cannot assign to this expression
  |
  = note: see issue #71126 <https://github.com/rust-lang/rust/issues/71126> for more information
```

### Version

```text
rustc 1.69.0-nightly (d962ea578 2023-02-26)
binary: rustc
commit-hash: d962ea57899d64dc8a57040142c6b498a57c8064
commit-date: 2023-02-26
host: aarch64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by dtolnay** [comment]
Mentioning @JirkaVebr @llogiq since this lint is from #10356.

**Comment by Jarcho** [comment]
Doesn't seem like a false positive. I'm not sure what the point of the suggestion to remove the `let` keyword is for since it goes against the stated reason for the lint.

**Comment by dtolnay** [comment]
> I'm not sure what the point of the suggestion to remove the `let` keyword is for

Based on the message outputted by the lint, I assumed this to be a "why waste time say lot word when few word do trick" lint: i.e. with the motivation that now that `_ = …` is available, prevent people from writing `let _ = …`, _except_ if they wanted to put a non-inferred type, which is not supported by `_ = …`.

> since it goes against the stated reason for the lint.

Yeah, going back and reading #6842, it seems my assumed interpretation is not at all the rationale behind the lint originally proposed. The purported reason that `let _ = …` is bad also applies equally to `_ = …`.

**Comment by JirkaVebr** [comment]
@dtolnay First of all, I'm really sorry about this issue. This was my first ever contribution made at @llogiq's excellent [Clippy workshop](https://users.rust-lang.org/t/rustnationuk-adjacent-clippy-workshop/88717), and I'm definitely no expert on this.

The intention behind the lint wasn't primarily to save tokens by dropping the `let`, but rather to nudge people towards providing a type ascription, thereby explicitly stating what exactly is it that they wish to ignore. That way their compile would break should the type of the ignored expression change, and they would be forced to re-evaluate the decision to ignore the value. Perhaps once the type is different, they would no longer wish to ignore the value? Hence the name of the lint: `let_underscore_untyped` ‒ it's really about the type. The "dropping the `let`" part is intended primarily for cases where the type cannot be named, is too complex, or the author just doesn't wish to spell out the type. But yes, `_ = <expr>` does suffer from the same problem the lint is trying to address.

> Based on the message outputted by the lint

Is the message too misleading then? Perhaps it should be more clear that the recommended resolution is to provide the type ascription?

Regarding the MSRV, that is just a bug and an oversight on our part, thanks for pointing that out. For older versions of Rust, should it not recommend dropping the `let` or should it just be disabled? :thinking:

**Comment by dtolnay** [comment]
IMO clippy should not ever suggest replacing `let _ = ...` with `_ = ...`, regardless of msrv, since that is not better.

**Comment by smoelius** [comment]
> This might be better as a `restriction` lint. `let _ = ...` is used very frequently and there's nothing really wrong with it.

I also think this might be better as a restriction lint.

**Comment by ryanavella** [comment]
I can think of one advantage to `_ = ...`, which is that it more closely follows my intuition of drop-order.

Normally `let` bindings keep a value alive until the end of scope, but `let _ = ...` breaks that pattern. This is a common enough footgun that newcomers will often ask why `let _ = lock.lock().unwrap();` doesn't lock a mutex for the duration of a scope, for example.

On the other hand, when I see a destructuring assignment like `(x, _) = (1, FooWithNonTrivialDrop);`, I expect `FooWithNonTrivialDrop` to be dropped immediately because it is not being bound to anything in particular. But I would like to know if this is consistent with other dev's intuitions, or if it is just peculiar to me and my own mental model.

**Comment by dtolnay** [comment]
I don't agree @ryanavella -- overall I don't feel that `_ =` puts anyone in a  better spot to get away not understanding the special behavior of underscore patterns, compared to `let _ =`. Whether the mental model is "`_ =` means drop immediately", or "`_ =` works like `ident =`", those are both wrong and `_` is its own thing, both in `_ =` and `let _ =`.

**Comment by JirkaVebr** [comment]
I think the fundamental problem really are types that cannot be named (#10411). If I recall correctly, @llogiq said that it is either incredibly difficult or even impossible for Clippy to identify these, and so the suggestion to drop the `let` came to be largely as a way for the lint to move out of the way so that people aren't forced to write `#[allow(clippy::let_underscore_untyped)]` when it's really not their fault.

I do fully agree that `_ = <expr>` has plenty of downsides though. Assuming it was feasible to detect unnameable types, the lint could only trigger when there's a type for the user to write, and it could drop the suggestion to omit the `let`. Otherwise, if it isn't feasible, is it acceptable for there to be edge cases where `#[allow(clippy::let_underscore_untyped)]` is necessary even though it was not possible to name the type?

Separately, the MSRV and #10441 are just bugs that could be fixed right away no matter the answer to these questions.

I would love to contribute a fix but I was hoping to first also hear @llogiq's thoughts on all this. Which path could/should the fix take? Can we recognize unnameable types?
